sdk/src/main/scala/com/horizen/SidechainApp.scala

```diff
@@ -203,16 +207,27 @@ class SidechainApp @Inject()
   // Generate snark keys only if were not present before.
   if (!Files.exists(Paths.get(params.certVerificationKeyFilePath)) || !Files.exists(Paths.get(params.certProvingKeyFilePath))) {
     log.info("Generating Cert snark keys. It may take some time.")
-    if (!CryptoLibProvider.sigProofThresholdCircuitFunctions.generateCoboundaryMarlinSnarkKeys(
-        sidechainSettings.withdrawalEpochCertificateSettings.maxPks, params.certProvingKeyFilePath, params.certVerificationKeyFilePath)) {
+    val expectedNumOfCustomFields = if (params.isCSWEnabled) CommonCircuit.customFieldsNumber else CommonCircuit.customFieldsNumberWithDisabledCSW
+    if (!CryptoLibProvider.sigProofThresholdCircuitFunctions.generateCoboundaryMarlinSnarkKeys(sidechainSettings.withdrawalEpochCertificateSettings.maxPks, params.certProvingKeyFilePath, params.certVerificationKeyFilePath, expectedNumOfCustomFields)) {
       throw new IllegalArgumentException("Can't generate Cert Coboundary Marlin ProvingSystem snark keys.")
     }
   }
-  if (!Files.exists(Paths.get(params.cswVerificationKeyFilePath)) || !Files.exists(Paths.get(params.cswProvingKeyFilePath))) {
-    log.info("Generating CSW snark keys. It may take some time.")
-    if (!CryptoLibProvider.cswCircuitFunctions.generateCoboundaryMarlinSnarkKeys(
-      params.withdrawalEpochLength, params.cswProvingKeyFilePath, params.cswVerificationKeyFilePath)) {
-      throw new IllegalArgumentException("Can't generate CSW Coboundary Marlin ProvingSystem snark keys.")
+  if (isCSWEnabled) {
+    if (Option(params.cswVerificationKeyFilePath).forall(_.trim.isEmpty)){
```

Comment:
I have looked this expression on Internet because I wanted a Scala-way to check if params.cswVerificationKeyFilePath was null or empty. If I remove the Option/forall part, if params.cswVerificationKeyFilePath is null there will be a NullPointerException. I don't know if params.cswVerificationKeyFilePath can be null, to be honest, because is read from the .conf file, so maybe it's too much, but if exists the possibility that it is null, I would prefer to keep the check.

Reply:
ok, fine